# Assignment - Toronto data from Wiki page
*Submitted by Vikram Seshadri for Coursera Capstone Project Assignment Week 3*

#### First step is to import the necessary libraries to extract information from the *"List_of_postal_codes_of_Canada:_M"* wikipedia page.

In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np

In [2]:
wiki_url = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text
soup = BeautifulSoup(wiki_url)

In [3]:
wiki_table = soup.find("table",class_ = "wikitable sortable")
PostalCode = []
Borough = []
Neighborhood = []
ColumnTitles = []
titles = soup.findAll('th')
for i in range(0,3):
    ColumnTitles.append(titles[i].find(text=True).strip('\n'))

for row in soup.findAll("tr"):
    cells = row.findAll('td')
    if len(cells)==3: #Only extract table body not heading
        PostalCode.append(str(cells[0].find(text=True).strip('\n')))
        Borough.append(str(cells[1].find(text=True).strip('\n')))
        Neighborhood.append(str(cells[2].find(text=True).strip('\n')))      

#### Convert the lists into a dataframe using Pandas

In [4]:
Toronto_df = pd.DataFrame(columns=ColumnTitles)
Toronto_df['Postcode'] = PostalCode
Toronto_df['Borough'] = Borough
Toronto_df['Neighbourhood'] = Neighborhood
Toronto_df.rename(columns={"Neighbourhood" : "Neighborhood"},inplace = True)
Toronto_df.head()

Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

#### Remove the rows that does not have an assigned Borough

In [5]:
Toronto_new_df = Toronto_df[Toronto_df.Borough != 'Not assigned']

In [6]:
Toronto_new_df.head(10)

Postcode           Borough      Neighborhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Not assigned
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge
12      M1B       Scarborough           Malvern

#### If a neighborhood is not assigned, assign the same name as that of the borough.

In [7]:
pd.set_option('mode.chained_assignment', None)
Toronto_new_df.Neighborhood = Toronto_new_df.Borough.where(Toronto_new_df.Neighborhood == 'Not assigned', Toronto_new_df.Neighborhood)
Toronto_new_df.head(10)

Postcode           Borough      Neighborhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Queen's Park
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge
12      M1B       Scarborough           Malvern

In [8]:
Toronto_new_df.shape

(211, 3)

#### Download the latitude and longitude locations for the corresponding postal codes from the *cocl* website.

In [9]:
postal_codes = Toronto_new_df['Postcode'].values.tolist()
LatLng = pd.read_csv('https://cocl.us/Geospatial_data')
LatLng.rename(columns={"Postal Code" : "Postcode"},inplace = True)
LatLng.head()

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

In [10]:
Toronto_loc_df = Toronto_new_df.join(LatLng.set_index('Postcode'), on='Postcode')
Toronto_disp_df = Toronto_loc_df
Toronto_postal_agg = pd.DataFrame(Toronto_disp_df.groupby(['Postcode','Borough'])['Neighborhood'].apply(lambda x: ','.join(x)))
Toronto_postal_disp = Toronto_postal_agg.join(LatLng.set_index('Postcode'), on='Postcode')
Toronto_postal_disp.reset_index(inplace=True)

#### Look up the names of unique boroughs to identify the names of boroughs that have the word Toronto in it.

In [11]:
Toronto_postal_disp.head(10)

Postcode      Borough                                   Neighborhood  \
0      M1B  Scarborough                                  Rouge,Malvern   
1      M1C  Scarborough           Highland Creek,Rouge Hill,Port Union   
2      M1E  Scarborough                Guildwood,Morningside,West Hill   
3      M1G  Scarborough                                         Woburn   
4      M1H  Scarborough                                      Cedarbrae   
5      M1J  Scarborough                            Scarborough Village   
6      M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park   
7      M1L  Scarborough                  Clairlea,Golden Mile,Oakridge   
8      M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West   
9      M1N  Scarborough                     Birch Cliff,Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

In [12]:
Toronto_postal_disp.shape

(103, 5)

#### Store the location details and Neighborhood details of the Boroughs that have the word Toronto in it.

In [13]:
Toronto_nbh = Toronto_postal_disp[Toronto_postal_disp['Borough'].str.contains('Toronto')].reset_index(drop=True)
Toronto_nbh.head(10)

Postcode          Borough  \
0      M4E     East Toronto   
1      M4K     East Toronto   
2      M4L     East Toronto   
3      M4M     East Toronto   
4      M4N  Central Toronto   
5      M4P  Central Toronto   
6      M4R  Central Toronto   
7      M4S  Central Toronto   
8      M4T  Central Toronto   
9      M4V  Central Toronto   

                                        Neighborhood   Latitude  Longitude  
0                                        The Beaches  43.676357 -79.293031  
1                        The Danforth West,Riverdale  43.679557 -79.352188  
2                      The Beaches West,India Bazaar  43.668999 -79.315572  
3                                    Studio District  43.659526 -79.340923  
4                                      Lawrence Park  43.728020 -79.388790  
5                                   Davisville North  43.712751 -79.390197  
6                                 North Toronto West  43.715383 -79.405678  
7                                         Davisville  43.704324 -79.388790  
8                         Moore Park,Summerhill East  43.689574 -79.383160  
9  Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049

In [14]:
Toronto_nbh.shape

(38, 5)

In [15]:
import json # library to handle JSON files
from geopy.geocoders import Nominatim 
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans
import folium # map rendering library
print("Packages Installed!")

Packages Installed!


#### Obtain the Latitude and Longitude location of Toronto.

In [16]:
address = 'Toronto, Canada'
geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of  Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of  Toronto are 43.653963, -79.387207.


#### Using Folium, visualize the map of Toronto with the Boroughs containing the name Toronto in it.

In [17]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Toronto_nbh['Latitude'], Toronto_nbh['Longitude'], Toronto_nbh['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

#### Provide Foursquare client ID, Client Secret, Version, and the limit of the search.

In [18]:
CLIENT_ID = 'KFQB4CA4OB0VCSZFMGBYRF3CWIUU5TVCPRNQXU3QXYXZ4VPT' # your Foursquare ID
CLIENT_SECRET = '2JP0S2ZVUZHSAXVJS53UYQOKZ0M1UE5SWDQ2GAMAA34EKNOC' # your Foursquare Secret
VERSION = '20190715' # Foursquare API version
LIMIT = 100

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: KFQB4CA4OB0VCSZFMGBYRF3CWIUU5TVCPRNQXU3QXYXZ4VPT
CLIENT_SECRET:2JP0S2ZVUZHSAXVJS53UYQOKZ0M1UE5SWDQ2GAMAA34EKNOC


#### Function that will obtain the nearby venues for any number of places with their corresponding latitudes and longitudes. The function returns a dataframe that has both the neighborhood location, and venue location and category. This function can be used subsequently to obtain the details of the venues near the neighborhoods in Toronto.

In [19]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):

        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [20]:
Toronto_nbh.columns

Index(['Postcode', 'Borough', 'Neighborhood', 'Latitude', 'Longitude'], dtype='object')

In [21]:
Toronto_venues = getNearbyVenues(names=Toronto_nbh['Neighborhood'],
                                   latitudes=Toronto_nbh['Latitude'],
                                   longitudes=Toronto_nbh['Longitude']
                                  )

In [22]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1707, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  The Beaches              43.676357              -79.293031   
1                  The Beaches              43.676357              -79.293031   
2                  The Beaches              43.676357              -79.293031   
3                  The Beaches              43.676357              -79.293031   
4  The Danforth West,Riverdale              43.679557              -79.352188   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                            Pantheon       43.677621       -79.351434   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4   Greek Restaurant

In [23]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            56   
Brockton,Exhibition Place,Parkdale Village                             22   
Business Reply Mail Processing Centre 969 Eastern                      19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     17   
Cabbagetown,St. James Town                                             45   
Central Bay Street                                                     86   
Chinatown,Grange Park,Kensington Market                               100   
Christie                                                               16   
Church and Wellesley                                                   88   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             35   
Davisville North                                                        7   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     16   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            18   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      4   
Harbord,University of Toronto                                          34   
Harbourfront East,Toronto Islands,Union Station                       100   
Harbourfront,Regent Park                                               50   
High Park,The Junction South                                           22   
Lawrence Park                                                           3   
Little Portugal,Trinity                                                64   
Moore Park,Summerhill East                                              4   
North Toronto West                                                     18   
Parkdale,Roncesvalles                                                  15   
Rosedale                                                                5   
Roselawn                                                                1   
Runnymede,Swansea                                                      37   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        41   
The Annex,North Midtown,Yorkville                                      24   
The Beaches                                                             4   
The Beaches West,India Bazaar                                          21   
The Danforth West,Riverdale                                            41   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             56   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      17   
Cabbagetown,St. James Town                                              45   
Central Bay Street                                                      86   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                16   
Church and Wellesley                                                    88   
Commerce Court,Victoria Hotel      

In [24]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 236 uniques categories.


#### Converting the venue category into numerical dummy values for performing K-Means algorithm.

In [25]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")
Toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Thrift / Vintage Store  Toy / Game Store  \
0             0         0  ...                       0                 0   
1             0         0  ...                       0                 0   
2             0         0  ...                       0                 0   
3             0         0  ...                       0                 0   
4             0         0  ...                       0                 0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      1              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wings Joint  Yoga Studio  
0                      0         0            0            0  
1                      0         0            0            0  
2                      0         0            0            0  
3                      0         0            0            0  
4                      0         0            0            0  

[5 rows x 236 columns]

#### One of the columns of the dataframe *Toronto_onehot*  already has the name *Neighborhood*. Therefore when adding the neighborhood information to this dataframe, the column is named as *Neighborhoods*. Note that there were 236 rows before this point and it increased to 237 after the Neighborhood addition.

In [26]:
Toronto_onehot.insert(0,'Neighborhoods', Toronto_venues['Neighborhood'], True)

In [27]:
Toronto_onehot.shape

(1707, 237)

#### Group the dataframe by neighborhood and take the mean of all the frequencies of occurence.

In [28]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhoods').mean().reset_index()
Toronto_grouped.head()

Neighborhoods  Afghan Restaurant  \
0                             Adelaide,King,Richmond                0.0   
1                                        Berczy Park                0.0   
2         Brockton,Exhibition Place,Parkdale Village                0.0   
3  Business Reply Mail Processing Centre 969 Eastern                0.0   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0  0.000000            0.000000      0.000000        0.000000   
1  0.000000            0.000000      0.000000        0.000000   
2  0.000000            0.000000      0.000000        0.000000   
3  0.000000            0.000000      0.000000        0.000000   
4  0.058824            0.058824      0.058824        0.117647   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0         0.000000          0.000000                 0.03           0.0  ...   
1         0.000000          0.000000                 0.00           0.0  ...   
2         0.000000          0.000000                 0.00           0.0  ...   
3         0.000000          0.000000                 0.00           0.0  ...   
4         0.176471          0.117647                 0.00           0.0  ...   

   Thrift / Vintage Store  Toy / Game Store  Trail  Train Station  \
0                     0.0               0.0    0.0            0.0   
1                     0.0               0.0    0.0            0.0   
2                     0.0               0.0    0.0            0.0   
3                     0.0               0.0    0.0            0.0   
4                     0.0               0.0    0.0            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                       0.010000               0.0                    0.0   
1                       0.017857               0.0                    0.0   
2                       0.000000               0.0                    0.0   
3                       0.000000               0.0                    0.0   
4                       0.000000               0.0                    0.0   

   Wine Bar  Wings Joint  Yoga Studio  
0      0.01          0.0     0.000000  
1      0.00          0.0     0.000000  
2      0.00          0.0     0.000000  
3      0.00          0.0     0.052632  
4      0.00          0.0     0.000000  

[5 rows x 237 columns]

In [29]:
Toronto_grouped.shape

(38, 237)

#### Display the top five venues for each neighborhood.

In [30]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhoods']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhoods'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
         venue  freq
0  Coffee Shop  0.07
1         Café  0.05
2          Bar  0.04
3   Steakhouse  0.04
4        Hotel  0.03


----Berczy Park----
                venue  freq
0         Coffee Shop  0.11
1        Cocktail Bar  0.05
2      Farmers Market  0.04
3         Cheese Shop  0.04
4  Seafood Restaurant  0.04


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0     Coffee Shop  0.09
1  Breakfast Spot  0.09
2            Café  0.09
3      Restaurant  0.05
4         Stadium  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1       Auto Workshop  0.05
2                 Spa  0.05
3       Burrito Place  0.05
4             Butcher  0.05


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3            

Function to sort the required number of venues in decending order of freqency.

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Use the above written function to obtain the top 10 venues from each neighborhood.

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhoods']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhoods'] = Toronto_grouped['Neighborhoods']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhoods 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                  Café                   Bar             Steakhouse   
1          Cocktail Bar              Beer Bar         Farmers Market   
2                  Café           Coffee Shop  Performing Arts Venue   
3           Yoga Studio         Auto Workshop            Pizza Place   
4        Airport Lounge      Airport Terminal                  Plane   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant                   Gym            Restaurant   
1           Cheese Shop    Seafood Restaurant                  Café   
2          Intersection         Burrito Place     Convenience Store   
3      Recording Studio            Restaurant               Butcher   
4       Harbor / Marina           Coffee Shop      Sculpture Garden   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Hotel          Burger Joint        Thai Restaurant  
1                Bakery            Steakhouse                 Museum  
2               Stadium  Caribbean Restaurant             Restaurant  
3         Burrito Place            Skate Park                Brewery  
4         Boat or Ferry              Boutique                    Bar

#### In this study, 7 clusters are chosen in order to get a better classification of Toronto neighborhoods

In [33]:
kclusters = 7

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhoods', 1)
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [34]:
len(kmeans.labels_)

38

In [35]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted

Cluster Labels                                      Neighborhoods  \
0                1                             Adelaide,King,Richmond   
1                1                                        Berczy Park   
2                1         Brockton,Exhibition Place,Parkdale Village   
3                1  Business Reply Mail Processing Centre 969 Eastern   
4                1  CN Tower,Bathurst Quay,Island airport,Harbourf...   
5                1                         Cabbagetown,St. James Town   
6                1                                 Central Bay Street   
7                1            Chinatown,Grange Park,Kensington Market   
8                1                                           Christie   
9                1                               Church and Wellesley   
10               1                      Commerce Court,Victoria Hotel   
11               1                                         Davisville   
12               0                                   Davisville North   
13               1  Deer Park,Forest Hill SE,Rathnelly,South Hill,...   
14               1            Design Exchange,Toronto Dominion Centre   
15               1                        Dovercourt Village,Dufferin   
16               1              First Canadian Place,Underground city   
17               6                 Forest Hill North,Forest Hill West   
18               1                      Harbord,University of Toronto   
19               1    Harbourfront East,Toronto Islands,Union Station   
20               1                           Harbourfront,Regent Park   
21               1                       High Park,The Junction South   
22               3                                      Lawrence Park   
23               1                            Little Portugal,Trinity   
24               4                         Moore Park,Summerhill East   
25               1                                 North Toronto West   
26               1                              Parkdale,Roncesvalles   
27               4                                           Rosedale   
28               2                                           Roselawn   
29               1                                  Runnymede,Swansea   
30               1                            Ryerson,Garden District   
31               1                                     St. James Town   
32               1                    Stn A PO Boxes 25 The Esplanade   
33               1                                    Studio District   
34               1                  The Annex,North Midtown,Yorkville   
35               5                                        The Beaches   
36               0                      The Beaches West,India Bazaar   
37               1                        The Danforth West,Riverdale   

   1st Most Common Venue          2nd Most Common Venue 3rd Most Common Venue  \
0            Coffee Shop                           Café                   Bar   
1            Coffee Shop                   Cocktail Bar              Beer Bar   
2         Breakfast Spot                           Café           Coffee Shop   
3     Light Rail Station                    Yoga Studio         Auto Workshop   
4        Airport Service                 Airport Lounge      Airport Terminal   
5            Coffee Shop                           Café            Restaurant   
6            Coffee Shop                 Sandwich Place                  Café   
7                   Café  Vegetarian / Vegan Restaurant    Chinese Restaurant   
8          Grocery Store                           Café                  Park   
9            Coffee Shop                        Gay Bar   Japanese Restaurant   
10           Coffee Shop                          Hotel                  Café   
11           Coffee Shop                 Sandwich Place          Dessert Shop   
12     Food & Drink Shop                 Breakfast Spot                   Gym   
13           Coffee Sh

#### Before merging the column *Neighborhood* from *neighborhood_venues_sorted*, rename the column *Neighborhood* to *Neighborhoods* in Toronto_merged dataframe.

In [36]:
Toronto_merged = Toronto_nbh
Toronto_merged.rename(columns = {'Neighborhood':'Neighborhoods'},inplace=True)
Toronto_merged.head()

Postcode          Borough                  Neighborhoods   Latitude  \
0      M4E     East Toronto                    The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3      M4M     East Toronto                Studio District  43.659526   
4      M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [37]:
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhoods'), on='Neighborhoods')
Toronto_merged.head() # check the last columns!

Postcode          Borough                  Neighborhoods   Latitude  \
0      M4E     East Toronto                    The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3      M4M     East Toronto                Studio District  43.659526   
4      M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               5     Health Food Store                   Pub   
1 -79.352188               1      Greek Restaurant           Coffee Shop   
2 -79.315572               0                  Park        Sandwich Place   
3 -79.340923               1                  Café           Coffee Shop   
4 -79.388790               3                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                 Trail          Neighborhood             Yoga Studio   
1    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
2          Burger Joint     Food & Drink Shop              Steakhouse   
3             Gastropub    Italian Restaurant                  Bakery   
4              Bus Line           Yoga Studio                 Dog Run   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   Dumpling Restaurant              Dive Bar               Dog Run   
1           Yoga Studio           Pizza Place             Bookstore   
2     Fish & Chips Shop  Fast Food Restaurant            Board Shop   
3   American Restaurant           Yoga Studio                  Park   
4  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant             Donut Shop  
1               Brewery        Bubble Tea Shop  
2               Brewery          Burrito Place  
3               Brewery     Seafood Restaurant  
4           Event Space   Ethiopian Restaurant

In [38]:
Toronto_merged.shape

(38, 16)

#### Folium map with locations colored corresponding to different clusters.

In [39]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)+5))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhoods'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Cluster 1:

In [40]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2     East Toronto               0                  Park   
5  Central Toronto               0     Food & Drink Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2        Sandwich Place          Burger Joint     Food & Drink Shop   
5        Breakfast Spot                   Gym                 Hotel   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2            Steakhouse     Fish & Chips Shop  Fast Food Restaurant   
5                  Park        Sandwich Place        Clothing Store   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2                   Board Shop               Brewery          Burrito Place  
5  Eastern European Restaurant               Dog Run       Doner Restaurant

#### Cluster 2:
* Contains Coffee shop and cafe in top 5
* All Boroughs have some common venues in them

In [41]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1       East Toronto               1      Greek Restaurant   
3       East Toronto               1                  Café   
6    Central Toronto               1           Coffee Shop   
7    Central Toronto               1           Coffee Shop   
9    Central Toronto               1           Coffee Shop   
11  Downtown Toronto               1           Coffee Shop   
12  Downtown Toronto               1           Coffee Shop   
13  Downtown Toronto               1           Coffee Shop   
14  Downtown Toronto               1           Coffee Shop   
15  Downtown Toronto               1           Coffee Shop   
16  Downtown Toronto               1           Coffee Shop   
17  Downtown Toronto               1           Coffee Shop   
18  Downtown Toronto               1           Coffee Shop   
19  Downtown Toronto               1           Coffee Shop   
20  Downtown Toronto               1           Coffee Shop   
21  Downtown Toronto               1           Coffee Shop   
24   Central Toronto               1           Coffee Shop   
25  Downtown Toronto               1                  Café   
26  Downtown Toronto               1                  Café   
27  Downtown Toronto               1       Airport Service   
28  Downtown Toronto               1           Coffee Shop   
29  Downtown Toronto               1           Coffee Shop   
30  Downtown Toronto               1         Grocery Store   
31      West Toronto               1              Pharmacy   
32      West Toronto               1                   Bar   
33      West Toronto               1        Breakfast Spot   
34      West Toronto               1                  Café   
35      West Toronto               1        Breakfast Spot   
36      West Toronto               1           Pizza Place   
37      East Toronto               1    Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue  \
1                     Coffee Shop    Italian Restaurant   
3                     Coffee Shop             Gastropub   
6                  Clothing Store           Yoga Studio   
7                  Sandwich Place          Dessert Shop   
9                             Pub    Light Rail Station   
11                           Café            Restaurant   
12                        Gay Bar   Japanese Restaurant   
13                         Bakery                  Park   
14                 Clothing Store        Cosmetics Shop   
15                           Café            Restaurant   
16                   Cocktail Bar              Beer Bar   
17                 Sandwich Place                  Café   
18                           Café                   Bar   
19                          Hotel              Aquarium   
20                           Café                 Hotel   
21                          Hotel                  Café   
24                 Sandwich Place                  Café   
25                            Bar   Japanese Restaurant   
26  Vegetarian / Vegan Restaurant    Chinese Restaurant   
27                 Airport Lounge      Airport Terminal   
28                           Café            Restaurant   
29                           Café            Restaurant   
30                           Café                  Park   
31                         Bakery           Supermarket   
32                    Coffee Shop      Asian Restaurant   
33                           Café           Coffee Shop   
34             Mexican Restaurant                   Bar   
35                      Gift Shop             Bookstore   
36                    Coffee Shop                  Café   
37                    Yoga Studio         Auto Workshop   

    4th Most Common Venue    5th Most Common Venue  6th Most Common Venue  \
1          Ice Cream Shop   Furniture / Home Store            Yoga Studio   
3      Italian Restaurant                   Bakery    American Restaurant   
6      Italian Restaurant    

#### Cluster 3:
* All of them are in Central Toronto

In [42]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               2                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22           Yoga Studio           Fish Market   Filipino Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22           Event Space  Ethiopian Restaurant      Electronics Store

#### Cluster 4:

In [43]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4  Central Toronto               3                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4           Swim School              Bus Line           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4               Dog Run  Fast Food Restaurant        Farmers Market   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4    Falafel Restaurant           Event Space   Ethiopian Restaurant

#### Cluster 5:
* Only one venue in central Toronto popular for Playground and Park

In [44]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8    Central Toronto               4            Playground   
10  Downtown Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8                   Park           Summer Camp          Tennis Court   
10            Playground                 Trail              Building   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8         Discount Store        Farmers Market    Falafel Restaurant   
10        Discount Store        Farmers Market    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8            Event Space  Ethiopian Restaurant      Electronics Store  
10           Event Space  Ethiopian Restaurant      Electronics Store

#### Cluster 6:
* Location in East Toronto popular for Health Food.

In [45]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 5, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  East Toronto               5     Health Food Store                   Pub   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail          Neighborhood           Yoga Studio   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   Dumpling Restaurant              Dive Bar               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant             Donut Shop

#### Cluster 7:

In [46]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 6, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
23  Central Toronto               6    Mexican Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23                 Trail      Sushi Restaurant         Jewelry Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23           Yoga Studio               Dog Run  Fast Food Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23        Farmers Market    Falafel Restaurant            Event Space

*This marks the end of this Jupyter Notebook that analyzes the neighborhoods in Toronto, Canada using KMeans Algorithm.*